crates/pipeline-manager/src/runner/local_runner.rs

```diff
@@ -127,33 +117,162 @@ pub async fn fetch_binary_ref(
 ///
 /// In addition, it attempts to non-blocking kill the spawned process upon
 /// `drop` (which occurs when the owning automaton exits its run loop).
-pub struct ProcessRunner {
+///
+/// The log follow requests are replied to by a background thread:
+/// - If the pipeline has not yet started, it sends back rejections
+/// - If it has started, the thread adds new followers and catches them up,
+///   and sends the latest log lines as they come in to existing followers
+pub struct LocalRunner {
     pipeline_id: PipelineId,
-    pipeline_process: Option<Child>,
-    config: Arc<LocalRunnerConfig>,
+    config: LocalRunnerConfig,
+    process: Option<Child>,
+    log_terminate_sender: Sender<()>,
+    log_join_handle: JoinHandle<Receiver<Sender<LogMessage>>>,
 }
 
-impl Drop for ProcessRunner {
+impl Drop for LocalRunner {
     fn drop(&mut self) {
         // This shouldn't normally happen, as the runner should shutdown the
         // pipeline before destroying the automaton, but we make sure that the
         // pipeline process is still killed on error.  We use `start_kill`
         // to avoid blocking in `drop`.
-        self.pipeline_process.as_mut().map(|p| p.start_kill());
+        self.process.as_mut().map(|p| p.start_kill());
+
+        // Abort the log thread
+        self.log_join_handle.abort();
     }
 }
 
-impl ProcessRunner {
+impl LocalRunner {
     async fn kill_pipeline(&mut self) {
-        if let Some(mut p) = self.pipeline_process.take() {
+        if let Some(mut p) = self.process.take() {
             let _ = p.kill().await;
             let _ = p.wait().await;
         }
     }
+
+    /// Sets up a thread which listens to log follow requests and new incoming lines
+    /// from stdout/stderr. New followers are caught up and existing followers receive
+    /// the new stdout/stderr lines as they come in.
+    ///
+    /// Returns a sender to invoke termination of the thread and the corresponding join handle.
+    fn setup_log_operational(
+        pipeline_id: PipelineId,
+        stdout: ChildStdout,
+        stderr: ChildStderr,
+        mut log_follow_request_receiver: Receiver<Sender<LogMessage>>,
+    ) -> (Sender<()>, JoinHandle<Receiver<Sender<LogMessage>>>) {
+        let (terminate_sender, mut terminate_receiver) = mpsc::channel::<()>(10);
```

Comment:
why 10

Reply:
I've now replaced it with a oneshot channel, it is only for sending the termination message.